# cosin_code

This repository contains the R code developed and used in the paper:

Structured factorization for single-cell gene expression data

## cosin package

The R code uses the [cosin](https://github.com/giovannitoto/cosin) package which implements COunt data Structured INfinite factorization (COSIN). To install the package (in R):
```{r}
# if devtools is not installed yet:
# install.packages("devtools")
library(devtools)
install_github("giovannitoto/cosin")
```

## Contents

The repository includes the following files and folders:

- [simulations](https://github.com/giovannitoto/cosin_code/tree/main/simulations) contains files generated by `simulations_*` R files;

- [src/metrics.R](https://github.com/giovannitoto/cosin_code/tree/main/src/metrics.R) contains functions used to evaluate the models estimated on synthetic  data sets;

- [src/aglmpca.R](https://github.com/giovannitoto/cosin_code/tree/main/src/aglmpca.R) contains an implementation of GLM-PCA approximation proposed by Townes et al. (2019);

- [table_metrics.R](https://github.com/giovannitoto/cosin_code/tree/main/table_metrics.R) contains code for aggregating the evaluations of the models estimated on synthetic data sets;

- `simulations_*` are R files related to simulations; see *Simulations* section below.


## Simulation study

To illustrate the validity and generality of COSIN, we assess its performances through a
simulation study. As competitor, we use the generalized principal component analysis (or
glm-pca, Townes et al., 2019), representing the current state-of-the-art in dimension reduction for sRNA-seq data.

We considers three scenarios in which *eta* and *lambda* have different sparsity structures

#####  Data generation and COSIN estimation
`simulation_woNA.R` simulates the synthetic data sets without NAs and estimates COSIN with and without meta-covariates on them; `simulations_NA.R` simulates the synthetic data sets with NAs (25%) and estimates COSIN with and without meta-covariates on them.
We can launch the files from the command prompt as follows:
```
R --slave --no-restore --file=simulations_woNA.R
R --slave --no-restore --file=simulations_NA.R
```

##### COSIN evaluation
`simulations_NA_metrics.R` and `simulations_woNA_metrics.R` evaluate the synthetic data sets with and without NA values respectively.
Currently, the following evaluation metrics can be computed:

- metrics obtained comparing the original rank-one contributions and the contributions estimated by COSIN at each iteration:

  - Accuracy
  - F1
  - Mean Absolute Error (MAE)
  - Mean Square Error (MSE)
  - Precision
  - Root Mean Square Error (RMSE)
  - Sensitivity
  - Specificity


- metrics obtained comparing the original gene expression matrix and the matrix predicted by COSIN at each MCMC iteration:

  - Mean Absolute Deviation (MAD)
  - Mean Absolute Error (MAE)
  - Root Mean Square Error (RMSE)


We can launch the files from the command prompt as follows:
```
R --slave --no-restore --file=simulations_NA_metrics.R
R --slave --no-restore --file=simulations_woNA_metrics.R
```

For each scenario `<s>`, the metrics are grouped in a table in which each row corresponds to a synthetic data set and each column to an information related to the data set (`n`, `p`, `sigma`, `r`) or to an evaluation metric.
See [TABLES.md](https://github.com/giovannitoto/cosin_code/tree/main/TABLES.md) for the complete list of the columns and a description the evaluation metrics which are currently implemented.

The tables generated by `simulations_NA_metrics.R` are saved as `table_metrics_nometa_woNA_p<p>_s<s>.RDS`, while the ones generated by `simulations_woNA_metrics.R` as `table_metrics_nometa_NA_p<p>_s<s>.RDS`.


##### GLM-PCA estimation and evaluation
First, `simulations_glmpca.R` estimates GLM-PCA with *k*=2,...,8 latent factors on the synthetic data sets with NAs and selects as benchmark the specification characterized by the lowest out-of-sample MAE. Then, since the available implementation of GLM-PCA does not work on data with NAs, we implemented the GLM-PCA approximation proposed by Townes et al. (2019), i.e. principal component analysis on Pearson residuals of the null Poisson model (see [src/aglmpca.R](https://github.com/giovannitoto/cosin_code/tree/main/src/aglmpca.R)).

We can launch the file from the command prompt as follows:
```
R --slave --no-restore --file=simulations_glmpca.R
```

For each scenario `<s>`, the metrics are grouped in a table in which each row corresponds to a synthetic data set and each column to an information related to the data set (`n`, `p`, `sigma`, `r`, `na_prop`) or to an evaluation metric.
See [TABLES.md](https://github.com/giovannitoto/cosin_code/tree/main/TABLES.md) for the complete list of the columns and a description the evaluation metrics which are currently implemented.

The tables generated by `simulations_glmpca.R` are saved as `table_metrics_meta_s<s>.RDS` and `table_metrics_nometa_s<s>.RDS`.

##### Tables of metrics

All tables generated in the previous steps are saved in [simulations/metrics](https://github.com/giovannitoto/cosin_code/tree/main/simulations/metrics). `metrics_table.R` can be opened in *R* or *RStudio* to compute the aggregated evaluation metrics and explore the results.



## References

Townes, F. W., Hicks, S. C., Aryee, M. J., and Irizarry, R. A. (2019). Feature selection
and dimension reduction for single-cell rna-seq based on a multinomial model. *Genome
biology* **20**, 1â€“16.
